File: src/main/java/com/github/windchopper/tools/log/browser/AsyncRunner.java
package com.github.windchopper.tools.log.browser;

import javafx.beans.property.BooleanProperty;
import javafx.scene.Cursor;
import javafx.stage.Stage;

import javax.annotation.PreDestroy;
import javax.enterprise.context.ApplicationScoped;
import javax.inject.Named;
import java.util.Collection;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@ApplicationScoped @Named("AsyncRunner") public class AsyncRunner {

    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    @PreDestroy void destroying() {
        executor.shutdown();
    }

    void runAsync(Stage stage, Collection<BooleanProperty> actionDisableProperties, Runnable action) {
        executor.execute(() -> {
            stage.getScene().setCursor(Cursor.WAIT);
            actionDisableProperties.forEach(property -> property.set(true));

            try {
                action.run();
            } finally {
                actionDisableProperties.forEach(property -> property.set(false));
                stage.getScene().setCursor(Cursor.DEFAULT);
            }
        });
    }

}

File: src/main/java/com/github/windchopper/tools/log/browser/AsyncRunner.kt
package com.github.windchopper.tools.log.browser

import javafx.beans.property.BooleanProperty
import javafx.scene.Cursor
import javafx.stage.Stage
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import javax.enterprise.context.ApplicationScoped
import javax.inject.Named

@ApplicationScoped @Named("AsyncRunner") class AsyncRunner {

    fun runAsyncWithBusyPointer(stage: Stage, actionDisableProperties: Collection<BooleanProperty>, action: Runnable) {
        GlobalScope.launch {
            stage.scene.cursor = Cursor.WAIT

            actionDisableProperties.forEach {
                it.set(true)
            }

            try {
                action.run()
            } finally {
                stage.scene.cursor = Cursor.DEFAULT

                actionDisableProperties.forEach{
                    it.set(false)
                }
            }
        }
    }

    fun runAsync(actionDisableProperties: Collection<BooleanProperty>, action: Runnable) {
        GlobalScope.launch {
            actionDisableProperties.forEach {
                it.set(true)
            }

            try {
                action.run()
            } finally {
                actionDisableProperties.forEach {
                    it.set(false)
                }
            }
        }
    }

}